File: clients/android/app/src/main/java/com/syncbridge/android/util/FormatUtils.kt
package com.syncbridge.android.util

import java.time.Instant
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

fun relativeTime(iso: String): String {
    if (iso.isBlank()) return "just now"
    return try {
        val instant = Instant.parse(iso.replace(" ", "T").let {
            if (it.endsWith("Z") || it.contains("+")) it else "${it}Z"
        })
        val seconds = ChronoUnit.SECONDS.between(instant, Instant.now())
        when {
            seconds < 5 -> "just now"
            seconds < 60 -> "${seconds}s ago"
            seconds < 3600 -> "${seconds / 60}m ago"
            else -> "${seconds / 3600}h ago"
        }
    } catch (_: Exception) {
        try {
            val f = DateTimeFormatter.ISO_DATE_TIME
            val instant = java.time.ZonedDateTime.parse(iso, f).toInstant()
            val seconds = ChronoUnit.SECONDS.between(instant, Instant.now())
            if (seconds < 60) "${seconds}s ago" else "${seconds / 60}m ago"
        } catch (_: Exception) {
            "just now"
        }
    }
}

fun formatBytes(n: Long): String = when {
    n < 1024 -> "$n B"
    n < 1024 * 1024 -> "${"%.1f".format(n / 1024.0)} KB"
    else -> "${"%.1f".format(n / (1024.0 * 1024.0))} MB"
}

fun truncate(text: String, max: Int): String {
    val t = text.trim()
    return if (t.length > max) t.take(max) + "…" else t.ifEmpty { "(empty)" }
}

fun stripHtml(html: String): String =
    html.replace(Regex("<[^>]+>"), " ")
        .replace(Regex("&nbsp;"), " ")
        .replace(Regex("&amp;"), "&")
        .replace(Regex("&lt;"), "<")
        .replace(Regex("&gt;"), ">")
        .replace(Regex("\\s+"), " ")
        .trim()

fun clipboardDisplayText(content: String, max: Int = 500): String {
    val raw = content.trim()
    if (raw.isEmpty()) return "(empty)"
    val cleaned = if (raw.contains('<') && raw.contains('>')) stripHtml(raw) else raw
    return truncate(cleaned, max)
}

File: clients/ios/FormatUtils.swift
// FormatUtils.swift — Matches Android util/FormatUtils.kt

import Foundation

func relativeTime(_ iso: String) -> String {
    if iso.isEmpty { return "just now" }
    let normalized = iso.contains(" ") && !iso.contains("T")
        ? iso.replacingOccurrences(of: " ", with: "T") + (iso.hasSuffix("Z") ? "" : "Z")
        : iso
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    let date = formatter.date(from: normalized)
        ?? ISO8601DateFormatter().date(from: iso)
        ?? ISO8601DateFormatter().date(from: normalized)
    guard let date else { return "just now" }
    let seconds = Int(-date.timeIntervalSinceNow)
    if seconds < 5 { return "just now" }
    if seconds < 60 { return "\(seconds)s ago" }
    if seconds < 3600 { return "\(seconds / 60)m ago" }
    return "\(seconds / 3600)h ago"
}

func formatBytes(_ n: Int64) -> String {
    if n < 1024 { return "\(n) B" }
    if n < 1024 * 1024 { return String(format: "%.1f KB", Double(n) / 1024.0) }
    return String(format: "%.1f MB", Double(n) / (1024.0 * 1024.0))
}

func clipboardDisplayText(_ content: String, max: Int = 500) -> String {
    let raw = content.trimmingCharacters(in: .whitespacesAndNewlines)
    if raw.isEmpty { return "(empty)" }
    var cleaned = raw
    if raw.contains("<") && raw.contains(">") {
        cleaned = raw.replacingOccurrences(of: "<[^>]+>", with: " ", options: .regularExpression)
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
    if cleaned.count <= max { return cleaned }
    return String(cleaned.prefix(max)) + "…"
}

func isImageContentType(_ type: String) -> Bool {
    type.hasPrefix("image/")
}

func isTextMime(_ mime: String) -> Bool {
    mime.hasPrefix("text/") || mime == "application/json"
}

func canCopyFile(_ file: FileItem) -> Bool {
    if file.status != "ready" { return false }
    let ext = (file.name as NSString).pathExtension.lowercased()
    return file.mimeType.hasPrefix("image/")
        || isTextMime(file.mimeType)
        || ["txt", "md", "csv", "json", "log"].contains(ext)
}
